extensions/cms/src/cmsUtils.ts

```diff
@@ -157,49 +208,105 @@ export class CmsUtils {
 		let provider = await this.getCmsService();
-		return provider.removeRegisteredServer(ownerUri, relativePath, registeredServerName).then((result) => {
-			return result;
-		});
+		let result = await provider.removeRegisteredServer(ownerUri, relativePath, registeredServerName);
+		return result;
 	}
 
 	public async addServerGroup(groupName: string, groupDescription: string, relativePath: string, ownerUri: string): Promise<boolean> {
 		let provider = await this.getCmsService();
-		return provider.addServerGroup(ownerUri, relativePath, groupName, groupDescription).then((result) => {
-			return result;
-		});
+		let result = await provider.addServerGroup(ownerUri, relativePath, groupName, groupDescription);
+		return result;
 	}
 
 	public async removeServerGroup(groupName: string, relativePath: string, ownerUri: string): Promise<boolean> {
 		let provider = await this.getCmsService();
-		return provider.removeServerGroup(ownerUri, relativePath, groupName).then((result) => {
-			return result;
-		});
+		let result = await provider.removeServerGroup(ownerUri, relativePath, groupName);
+		return result;
 	}
 
 	// Getters
 	public get registeredCmsServers(): ICmsResourceNodeInfo[] {
 		return this._registeredCmsServers;
 	}
 
-	public get connection(): Thenable<azdata.connection.Connection> {
-		return this.openConnectionDialog([cmsProvider], undefined, { saveConnection: false }).then((connection) => {
+	public async credentialProvider(): Promise<azdata.CredentialProvider> {
+		if (!this._credentialProvider) {
+			this._credentialProvider = await azdata.credentials.getProvider(CredentialNamespace);
+		}
+		return this._credentialProvider;
+	}
+
+	public async connection(initialConnectionProfile?: azdata.IConnectionProfile): Promise<azdata.connection.Connection> {
+		if (!initialConnectionProfile) {
+			initialConnectionProfile = {
+				connectionName: undefined,
+				serverName: undefined,
+				databaseName: undefined,
+				userName: undefined,
+				password: undefined,
+				authenticationType: undefined,
+				savePassword: undefined,
+				groupFullName: undefined,
+				groupId: undefined,
+				providerName: undefined,
+				saveProfile: undefined,
+				id: undefined,
+				options: {}
+			};
+		}
+		try {
+			let connection = await this.openConnectionDialog([cmsProvider], initialConnectionProfile, { saveConnection: false });
 			if (connection) {
 				// remove group ID from connection if a user chose connection
 				// from the recent connections list
 				connection.options['groupId'] = null;
 				connection.providerName = mssqlProvider;
+				if (connection.options.savePassword) {
+					await this.savePassword(connection.options.user, connection.options.password);
+				}
 				return connection;
+			} else {
+				return Promise.reject();
 			}
-		});
+		} catch (error) {
+			throw error;
+		}
+	}
+
+	// Static Functions
+
+	public getConnectionProfile(connection: azdata.connection.Connection): azdata.IConnectionProfile {
```

Comment:
If you're removing the static they should be moved out of the class since they aren't using any of the state

Reply:
These are CMS related helper functions. I rather have them in the class.

Reply:
Fine, but at least make them static again since I don't see any of them using the service

---
In reply to: [295487169](https://github.com/microsoft/azuredatastudio/pull/5989#discussion_r295487169) [](ancestors = 295487169)